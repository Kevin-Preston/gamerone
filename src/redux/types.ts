import { AuthState } from './auth/types';
import { ProfileState } from './profile/types';
import { SettingsState } from './settings/types';
import { PostState } from './post/types';
import { PostFormState } from './post-form/types';
import { RequestStatusState } from './request-status/types';
import { RouterState } from 'connected-react-router';

export const INIT_STATE = 'root/INIT_STATE';
export const LOAD_STATE = 'root/LOAD_STATE';

export interface InitStateAction {
  type: typeof INIT_STATE;
}

export type RootStateActions = InitStateAction;

export interface RootState {
  Auth: AuthState;
  Profile: ProfileState;
  Settings: SettingsState;
  Post: PostState;
  PostForm: PostFormState;
  RequestStatus: RequestStatusState;
  router: RouterState;
}
