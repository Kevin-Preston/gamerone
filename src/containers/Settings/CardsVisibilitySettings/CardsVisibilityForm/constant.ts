export const VISIBLE_CARDS = [
  { key: 1, name: 'sponsors', label: 'Sponsors' },
  { key: 2, name: 'socialNetworks', label: 'Social Networks' },
  { key: 3, name: 'currentlyPlaying', label: 'Currently Playing' },
  { key: 4, name: 'gear', label: 'Gears' },
  { key: 5, name: 'achievements', label: 'Achievements' },
  { key: 6, name: 'store', label: 'Products' },
];
