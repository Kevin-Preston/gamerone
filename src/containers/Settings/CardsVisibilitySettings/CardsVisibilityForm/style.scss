.profile__cards--visibility .checkbox-label {
  max-width: 10rem;
}
