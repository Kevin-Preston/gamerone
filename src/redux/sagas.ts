import { all } from 'redux-saga/effects';
import authSagas from './auth/saga';
import profileSagas from './profile/saga';
import settingsSagas from './settings/saga';
import postSagas from './post/saga';
import postFormSagas from './post-form/saga';
import miscSagas from './misc/saga';

export default function* rootSaga() {
  yield all([
    authSagas(),
    profileSagas(),
    settingsSagas(),
    postSagas(),
    postFormSagas(),
    miscSagas(),
  ]);
}
