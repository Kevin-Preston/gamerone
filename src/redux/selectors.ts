/**
 * Here comes the combined selectors from different states
 */

import { createSelector } from 'reselect';

import { selectCurrentUser } from './auth/selectors';

import {
  selectCurrentProfileUser,
  selectCurrentProfileNetworks,
  selectCurrentProfileSponsors,
  selectCurrentProfileProducts,
} from './profile/selectors';

import {
  selectSettingsSponsors,
  selectSettingsProducts,
  selectSettingsProfileUser,
  selectProfileSocialNetworks,
} from './settings/selectors';

// determines whether the currently viewing profile is my profile or not
export const selectIsSelfProfile = createSelector(
  [selectCurrentProfileUser, selectCurrentUser],
  (profileUser, authUser) => authUser.id === profileUser?.id,
);

export const selectProfileUser = createSelector(
  [selectCurrentProfileUser, selectSettingsProfileUser],
  (profileUser, settingsUser) =>
    settingsUser.id === profileUser?.id ? settingsUser : profileUser,
);

// resolves social networks to be shown on Profile page
export const selectSocialNetworks = createSelector(
  [
    selectIsSelfProfile,
    selectProfileSocialNetworks,
    selectCurrentProfileNetworks,
  ],
  (isSelfProfile, settingsNetworks, profileNetworks) =>
    isSelfProfile ? settingsNetworks : profileNetworks,
);

// resolves sponsors to be shown on Profile page
export const selectSponsors = createSelector(
  [selectIsSelfProfile, selectSettingsSponsors, selectCurrentProfileSponsors],
  (isSelfProfile, settingsSponsors, profileSponsors) =>
    isSelfProfile ? settingsSponsors : profileSponsors,
);

// resolves products to be shown on Profile page
export const selectProducts = createSelector(
  [selectIsSelfProfile, selectSettingsProducts, selectCurrentProfileProducts],
  (isSelfProfile, settingsProducts, profileProducts) =>
    isSelfProfile ? settingsProducts : profileProducts,
);
