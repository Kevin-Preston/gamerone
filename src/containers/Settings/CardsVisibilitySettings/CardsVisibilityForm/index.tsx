import React, { useEffect } from 'react';
import { useForm } from 'react-hook-form';
import { useSelector, useDispatch } from 'react-redux';

import Button, { ButtonSchemeEnum } from 'components/common/Button';
import Checkbox from 'components/common/Form/Checkbox';

import { LayoutSettings } from 'interfaces';
import ProfileActions from 'redux/profile/actions';
import { selectSetLayoutStatus } from 'redux/request-status/selectors';
import { selectLayoutSettings } from 'redux/settings/selectors';

import { VISIBLE_CARDS as visibilityOptions } from './constant';
import './style.scss';

function CardsVisibilityForm() {
  const { register, handleSubmit, formState, reset } = useForm();
  const { dirty } = formState;

  const profileLayout = useSelector(selectLayoutSettings);
  const setStatus = useSelector(selectSetLayoutStatus);
  const dispatch = useDispatch();

  useEffect(() => {
    if (profileLayout?.visibility) {
      reset(profileLayout?.visibility);
    } else {
      reset();
    }
  }, [profileLayout, reset]);

  const onSubmit = (data: Record<string, any>) => {
    const layoutSettings: LayoutSettings = {
      settings: profileLayout?.settings,
      visibility: data,
    };
    dispatch(ProfileActions.setCurrentProfileLayout(layoutSettings));
  };

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <div className="profile__cards--visibility">
        {visibilityOptions.map((card) => (
          <Checkbox
            key={card.key}
            name={card.name}
            label={card.label}
            inputRef={register}
          />
        ))}
      </div>
      <div>
        <Button
          disabled={!dirty}
          scheme={ButtonSchemeEnum.PRIMARY}
          submitting={setStatus?.isFetching}
        >
          Save
        </Button>
      </div>
    </form>
  );
}

export default React.memo(CardsVisibilityForm);
