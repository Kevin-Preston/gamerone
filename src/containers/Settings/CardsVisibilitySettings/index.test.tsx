import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { render, fireEvent, cleanup } from '@testing-library/react';
import { act } from 'react-dom/test-utils';

import { store } from 'redux/store';
import CardsVisibilityForm from './index';

// Snapshot
test('cards visibility section initial', () => {
  const component = renderer.create(
    <Provider store={store}>
      <MemoryRouter>
        <CardsVisibilityForm />
      </MemoryRouter>
    </Provider>,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

// DOM testing
afterEach(cleanup);
it('check and save', () => {
  const { queryByLabelText, getByText } = render(
    <Provider store={store}>
      <MemoryRouter>
        <CardsVisibilityForm />
      </MemoryRouter>
    </Provider>,
  );

  expect(queryByLabelText(/sponsors/i)).toBeFalsy();
  expect(queryByLabelText(/gear/i)).toBeFalsy();

  act(() => {
    fireEvent.click(getByText('Save'));
  });

  expect(queryByLabelText(/gear/i)).toBeTruthy();
  expect(queryByLabelText(/sponsors/i)).toBeTruthy();
});

/**
 * Mock input event to do check
 *
 * @return {Object}
 */
const renderCardsVisibilityForm = () => {
  return render(
    <Provider store={store}>
      <MemoryRouter>
        <CardsVisibilityForm />
      </MemoryRouter>
    </Provider>,
  );
};

/**
 * Fill in and submit the form
 *
 * @param {Object} utils
 * @param {String} sponsors
 * @param {String} socialNetworks
 * @param {String} gear
 * @param {String} store
 * @param {String} currentlyPlaying
 * @param {String} achievements
 */
const fillAndSubmitForm = (
  utils,
  {
    sponsors = true,
    socialNetworks = false,
    gear = true,
    store = true,
    currentlyPlaying = false,
    achievements = true,
  },
) => {
  const { getByRole, getByLabelText } = utils;

  act(() => {
    fireEvent.change(getByLabelText(/sponsors/i), {
      target: { value: sponsors },
    });
    fireEvent.change(getByLabelText(/socialNetworks/i), {
      target: { value: socialNetworks },
    });
    fireEvent.change(getByLabelText(/currentlyPlaying/i), {
      target: { value: currentlyPlaying },
    });
    fireEvent.change(getByLabelText(/achievements/i), {
      target: { value: achievements },
    });
    fireEvent.change(getByLabelText(/gear/i), {
      target: { value: gear },
    });
    fireEvent.change(getByLabelText(/store/i), {
      target: { value: store },
    });
    fireEvent.submit(getByRole('button', { name: 'Save' }));
  });
};

test('update layoutsettings', async () => {
  const utils = renderCardsVisibilityForm();
  const visibility = {
    sponsors: true,
    socialNetworks: false,
    currentlyPlaying: true,
    achievements: true,
    gear: true,
    store: true,
  };

  // mock out window.fetch for the test
  const spy = jest.spyOn(window, 'fetch').mockImplementation((url) => {
    let success = {};
    if (url.endsWith('profile/layout-settings')) {
      success = visibility;
    }
    return Promise.resolve({
      status: 200,
      json: () => success,
    });
  });

  act(() => {
    fireEvent.click(utils.getByText('Save'));
  });

  fillAndSubmitForm(utils, visibility);

  spy.mockRestore();
});
