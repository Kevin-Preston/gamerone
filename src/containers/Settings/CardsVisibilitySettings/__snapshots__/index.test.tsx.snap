// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cards visibility section initial 1`] = `
<div
  className="card"
>
  <div
    className="card__content"
  >
    <h4>
      Cards
    </h4>
    <form
      onSubmit={[Function]}
    >
      <div
        className="profile__cards--visibility"
      >
        <div
          style={
            Object {
              "display": "flex",
              "width": "100%",
            }
          }
        >
          <label
            className="checkbox-label"
            htmlFor="sponsors"
          >
            Sponsors
          </label>
          <div
            className="checkbox-wrapper"
          >
            <input
              aria-label="sponsors"
              className="checkbox"
              id="sponsors"
              name="sponsors"
              type="checkbox"
            />
          </div>
        </div>
        <div
          style={
            Object {
              "display": "flex",
              "width": "100%",
            }
          }
        >
          <label
            className="checkbox-label"
            htmlFor="socialNetworks"
          >
            Social Networks
          </label>
          <div
            className="checkbox-wrapper"
          >
            <input
              aria-label="socialNetworks"
              className="checkbox"
              id="socialNetworks"
              name="socialNetworks"
              type="checkbox"
            />
          </div>
        </div>
        <div
          style={
            Object {
              "display": "flex",
              "width": "100%",
            }
          }
        >
          <label
            className="checkbox-label"
            htmlFor="currentlyPlaying"
          >
            Currently Playing
          </label>
          <div
            className="checkbox-wrapper"
          >
            <input
              aria-label="currentlyPlaying"
              className="checkbox"
              id="currentlyPlaying"
              name="currentlyPlaying"
              type="checkbox"
            />
          </div>
        </div>
        <div
          style={
            Object {
              "display": "flex",
              "width": "100%",
            }
          }
        >
          <label
            className="checkbox-label"
            htmlFor="gear"
          >
            Gears
          </label>
          <div
            className="checkbox-wrapper"
          >
            <input
              aria-label="gear"
              className="checkbox"
              id="gear"
              name="gear"
              type="checkbox"
            />
          </div>
        </div>
        <div
          style={
            Object {
              "display": "flex",
              "width": "100%",
            }
          }
        >
          <label
            className="checkbox-label"
            htmlFor="achievements"
          >
            Achievements
          </label>
          <div
            className="checkbox-wrapper"
          >
            <input
              aria-label="achievements"
              className="checkbox"
              id="achievements"
              name="achievements"
              type="checkbox"
            />
          </div>
        </div>
        <div
          style={
            Object {
              "display": "flex",
              "width": "100%",
            }
          }
        >
          <label
            className="checkbox-label"
            htmlFor="store"
          >
            Products
          </label>
          <div
            className="checkbox-wrapper"
          >
            <input
              aria-label="store"
              className="checkbox"
              id="store"
              name="store"
              type="checkbox"
            />
          </div>
        </div>
      </div>
      <div>
        <button
          className="button button--primary button--medium is-disabled"
          disabled={true}
          type="submit"
        >
          Save
        </button>
      </div>
    </form>
  </div>
</div>
`;
